exchange/utils.go

```diff
@@ -957,3 +959,49 @@ func WrapJSONInData(data []byte) []byte {
 	res = append(res, []byte(`}`)...)
 	return res
 }
+
+func getMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	mType := bid.MType
+	var bidType openrtb_ext.BidType
+	if mType > 0 {
+		switch mType {
+		case openrtb2.MarkupBanner:
+			bidType = openrtb_ext.BidTypeBanner
+		case openrtb2.MarkupVideo:
+			bidType = openrtb_ext.BidTypeVideo
+		case openrtb2.MarkupAudio:
+			bidType = openrtb_ext.BidTypeAudio
+		case openrtb2.MarkupNative:
+			bidType = openrtb_ext.BidTypeNative
+		default:
+			return bidType, fmt.Errorf("Failed to parse bid mType for impression \"%s\"", bid.ImpID)
+		}
+	} else {
+		var err error
+		bidType, err = getPrebidMediaTypeForBid(bid)
+		if err != nil {
+			return bidType, err
+		}
+	}
+	return bidType, nil
+}
+
+func getPrebidMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
+	if bid.Ext != nil {
+		var bidExt openrtb_ext.ExtBid
+		err := json.Unmarshal(bid.Ext, &bidExt)
+		if err == nil && bidExt.Prebid != nil {
+			bidType, bidTypeErr := openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
+			if bidTypeErr != nil {
```

Comment:
Per our offline discussion we agreed to append error from `ParseBidType` to result message. With this implementation if error occurs in `ParseBidType` it becomes shadowed and `if err != nill` check returns false and `err.Error()` concatenation never executes.
![Screenshot 2023-05-17 at 5 49 36 PM](https://github.com/prebid/prebid-server/assets/4790765/eda91bfe-a239-4152-89c1-83b0667dd1fb)


To fix this, 2 different error variables should be declared, here is a working version:

```go

func getPrebidMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
	var err, err2 error
	var bt openrtb_ext.BidType
	if bid.Ext != nil {
		var bidExt openrtb_ext.ExtBid
		err = json.Unmarshal(bid.Ext, &bidExt)
		if err == nil && bidExt.Prebid != nil {
			bt, err2 = openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
			if err2 == nil {
				return bt, nil
			}
		}
	}

	errMsg := fmt.Sprintf("Failed to parse bid mediatype for impression \"%s\"", bid.ImpID)
	if err2 != nil {
		errMsg = fmt.Sprintf("%s, %s", errMsg, err2.Error())
	}

	return "", &errortypes.BadServerResponse{
		Message: errMsg,
	}
}

``` 
It feels a little too complicated for such a simple function, but I can modify it if you think it's better.

Reply:
Technically at this point bid.ext is always correct. Bids are returned from adapters and in most cases adapters unmarshal json request to BidResponse hence incorrect bid.ext error will exist already. In case if bidder has it's own format and for whatever reason just set invalid byte array to bid.ext, this will be intercepted in `makeBidExtJSON`. 
But we need to handle this error properly anyways. Changed it to:

```go

func getPrebidMediaTypeForBid(bid openrtb2.Bid) (openrtb_ext.BidType, error) {
	var err error
	errFormat := "Failed to parse bid mediatype for impression \"%s\", %s"
	if bid.Ext != nil {
		var bidExt openrtb_ext.ExtBid
		err = json.Unmarshal(bid.Ext, &bidExt)
		if err == nil && bidExt.Prebid != nil {
			bidType, bidTypeErr := openrtb_ext.ParseBidType(string(bidExt.Prebid.Type))
			if bidTypeErr != nil {
				return "", &errortypes.BadServerResponse{
					Message: fmt.Sprintf(errFormat, bid.ImpID, bidTypeErr.Error()),
				}
			}
			return bidType, nil
		}
	}
	return "", &errortypes.BadServerResponse{
		Message: fmt.Sprintf(errFormat, bid.ImpID, err.Error()),
	}
}

```